core/src/test/java/hudson/os/WindowsUtil.java

```diff
@@ -0,0 +1,117 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package hudson.os;
+
+import hudson.Functions;
+import org.apache.commons.io.IOUtils;
+
+import javax.annotation.Nonnull;
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static org.junit.Assert.assertTrue;
+
+// adapted from:
+// https://blogs.msdn.microsoft.com/twistylittlepassagesallalike/2011/04/23/everyone-quotes-command-line-arguments-the-wrong-way/
+public class WindowsUtil {
+    private static final Pattern NEEDS_QUOTING = Pattern.compile("[\\s\"]");
+
+    /**
+     * Quotes an argument while escaping special characters interpreted by CreateProcess.
+     */
+    public static @Nonnull String quoteArgument(@Nonnull String argument) {
+        if (!NEEDS_QUOTING.matcher(argument).find()) return argument;
+        StringBuilder sb = new StringBuilder();
+        sb.append('"');
+        int end = argument.length();
+        for (int i = 0; i < end; i++) {
+            int nrBackslashes = 0;
+            while (i < end && argument.charAt(i) == '\\') {
+                i++;
+                nrBackslashes++;
+            }
+
+            if (i == end) {
+                // backslashes at the end of the argument must be escaped so the terminate quote isn't
+                nrBackslashes = nrBackslashes * 2;
+            } else if (argument.charAt(i) == '"') {
+                // backslashes preceding a quote all need to be escaped along with the quote
+                nrBackslashes = nrBackslashes * 2 + 1;
+            }
+            // else backslashes have no special meaning and don't need to be escaped here
+
+            for (int j = 0; j < nrBackslashes; j++) {
+                sb.append('\\');
+            }
+
+            if (i < end) {
+                sb.append(argument.charAt(i));
+            }
+        }
+        return sb.append('"').toString();
+    }
+
+    private static final Pattern CMD_METACHARS = Pattern.compile("[()%!^\"<>&|]");
+
+    /**
+     * Quotes an argument while escaping special characters suitable for use as an argument to {@code cmd.exe}.
+     */
+    public static @Nonnull String quoteArgumentForCmd(@Nonnull String argument) {
+        return CMD_METACHARS.matcher(quoteArgument(argument)).replaceAll("^$0");
+    }
+
+    /**
+     * Executes a command and arguments using {@code cmd.exe /C ...}.
+     */
+    public static @Nonnull Process execCmd(String... argv) throws IOException {
+        String command = Arrays.stream(argv).map(WindowsUtil::quoteArgumentForCmd).collect(Collectors.joining(" "));
+        return Runtime.getRuntime().exec(new String[]{"cmd.exe", "/C", command});
```

Comment:
this smells bad - you are passing in an array but then collapsing it to a single argument for the call to exex?  why not just keep the array - is there a reason (does cmd.exe loose that information)?

Reply:
The problem is that without escaping, there is no way to distinguish between whitespace in a file name and whitespace denoting different command line args. If we passed all the args properly to cmd.exe, cmd.exe will lose the quoting when it evals the command given. Windows is more annoying than bash in nested escaping.

Reply:
Heh, that link is in the code comments for this class.